compiler/rustc_ty_utils/src/layout.rs

```diff
@@ -1105,15 +1105,13 @@ fn variant_info_for_adt<'tcx>(
     };
 
     match layout.variants {
+        Variants::Empty => (vec![], None),
+
         Variants::Single { index } => {
-            if !adt_def.variants().is_empty() && layout.fields != FieldsShape::Primitive {
```

Comment:
What about the primitive case?

Reply:
I don't think that check was necessary. I don't know why it was added. This an ADT and ADTs are never primitive, except for variantless enums which use the `!` type.